easybuild/easyblocks/c/cudacompat.py

```diff
@@ -0,0 +1,263 @@
+        return self._has_nvidia_smi
+
+    def _run_nvidia_smi(self, args):
+        """
+        Run nvidia-smi with the given argument(s) and return the output.
+
+        Also does sensible logging.
+        Raises RuntimeError on failure.
+        """
+        if not self.has_nvidia_smi:
+            raise RuntimeError('Could not find nvidia-smi.')
+        cmd = 'nvidia-smi ' + args
+        out, ec = run_cmd(cmd, log_ok=False, log_all=False, regexp=False)
+        if ec != 0:
+            raise RuntimeError("`%s` returned exit code %s with output:\n%s" % (cmd, ec, out))
+        else:
+            self.log.info('`%s` succeeded with output:\n%s' % (cmd, out))
+            return out.strip().split('\n')
+
+    def prepare_step(self, *args, **kwargs):
+        """Parse and check the compatible_driver_versions value of the EasyConfig"""
+        compatible_driver_versions = self.cfg.get('compatible_driver_versions')
+        if compatible_driver_versions:
+            try:
+                # Create a dictionary with the major version as the keys
+                self.compatible_driver_version_map = {
+                    int(v.split('.', 1)[0]): v
+                    for v in compatible_driver_versions
+                }
+            except ValueError:
+                raise EasyBuildError("Invalid format of compatible_driver_versions. "
+                                     "Expected numeric major versions, got '%s'", compatible_driver_versions)
+        else:
+            self.compatible_driver_version_map = None
+        super(EB_CUDAcompat, self).prepare_step(*args, **kwargs)
+
+    def fetch_step(self, *args, **kwargs):
+        """Check for EULA acceptance prior to getting sources."""
+        # EULA for NVIDIA driver must be accepted via --accept-eula-for EasyBuild configuration option,
+        # or via 'accept_eula = True' in easyconfig file
+        self.check_accepted_eula(
+            name='NVIDIA-driver',
+            more_info='https://www.nvidia.com/content/DriverDownload-March2009/licence.php?lang=us'
+        )
+        return super(EB_CUDAcompat, self).fetch_step(*args, **kwargs)
+
+    def extract_step(self):
+        """Extract the files without running the installer."""
+        execpath = self.src[0]['path']
+        tmpdir = os.path.join(self.builddir, 'tmp')
+        targetdir = os.path.join(self.builddir, 'extracted')
+        run_cmd("/bin/sh " + execpath + " --extract-only --tmpdir='%s' --target '%s'" % (tmpdir, targetdir))
+        self.src[0]['finalpath'] = targetdir
+
+    def test_step(self):
+        """
+        Check for a compatible driver version if the EC has that information.
+
+        This can be skipped with `--skip-test-step`.
+        """
+
+        if self.compatible_driver_version_map and self.has_nvidia_smi:
+            try:
+                out_lines = self._run_nvidia_smi('--query-gpu=driver_version --format=csv,noheader')
```

Comment:
So an issue with this is that if someone is filtering `LD_LIBRARY_PATH` (like we do in EESSI) then this test will fail. One way I've done this in other places is to check for `self.installdir` in `LD_LIBRARY_PATH` and if not add the path. In this case you really do want that setting though, as you don't want to point to fixed path.

The other option is to actually catch the setting `EASYBUILD_FILTER_ENV_VARS=LD_LIBRARY_PATH` and say that this is not an option for this package. In general there are going to be issues with using this with an RPATH approach, it's solvable but the you need some funky configuration.

Reply:
I don't understand how this is related to the call of `nvidia-smi` at this point.

Also: What do you mean by "filtering `LD_LIBRARY_PATH`"?

>  In this case you really do want that setting though, as you don't want to point to fixed path.

Which setting?

> In general there are going to be issues with using this with an RPATH approach

Which issues can there be? `libcuda.so` etc are all loaded dynamically.

Reply:
Sorry I commented on the wrong line here, it should have been line 249, where you actually check that CUDA capability has changed.

> I don't understand how this is related to the call of `nvidia-smi` at this point.
> 
> Also: What do you mean by "filtering `LD_LIBRARY_PATH`"?

So there is an option in EasyBuild to filter out environment variables from the final module file. It's common enough if using rpath to do that for `LD_LIBRARY_PATH`.

> 
> > In this case you really do want that setting though, as you don't want to point to fixed path.
> 
> Which setting?

Here I mean you _don't_ want an RPATH to `libcuda.so`, you want to force the usage of `LD_LIBRARY_PATH` in the module file. Otherwise, this would make it impossible for you to use the `.modulerc` files to control which `libcuda.so` is actually used. 

> 
> > In general there are going to be issues with using this with an RPATH approach
> 
> Which issues can there be? `libcuda.so` etc are all loaded dynamically.

If someone used the EasyBuild rpath settings, the compat library path will be injected into executables that use it. This would remove your ability to dynamically change the `libcuda.so` that is actually used.

Reply:
Moving comment to correct location at https://github.com/easybuilders/easybuild-easyblocks/pull/2764/files#r996876865